cmd/publicize/server.go

```diff
@@ -0,0 +1,154 @@
+		s.createComment(ic, fmt.Sprintf("Publicize plugin can only be used in pull requests"), logger)
+		return
+	}
+
+	// Only org members should be able to publicize a pull request.
+	ok, err := s.ghc.IsMember(org, commentAuthor)
+	if err != nil {
+		return
+	}
+	if !ok {
+		s.createComment(ic, fmt.Sprintf("only [%s](https://github.com/orgs/%s/people) org members may request a publicize", org, org), logger)
+		return
+	}
+	logger.Info("Publicize of PR has been requested.")
+
+	pr, err := s.ghc.GetPullRequest(org, repo, num)
+	if err != nil {
+		return
+	}
+	baseBranch := pr.Base.Ref
+
+	if !pr.Merged {
+		s.createComment(ic, "cannot publicize an unmerged PR", logger)
+		return
+	}
+
+	logger.Info("Searching bond...")
+	bondedOrgRepo, ok := s.bonds[fmt.Sprintf("%s/%s", org, repo)]
+	if !ok {
+		logger.Warn("There is no public repository configured for the current repository.")
+		s.createComment(ic, "cannot publicize because there is no bond for the current pull request", logger)
+		return
+	}
+
+	bondedOrg := strings.Split(bondedOrgRepo, "/")[0]
+	bondedRepo := strings.Split(bondedOrgRepo, "/")[1]
+
+	logger.Infof("Trying to merge the PR to destination: %s/%s@%s", bondedOrg, bondedRepo, baseBranch)
+	if err := s.mergeAndPushToRemote(bondedOrg, bondedRepo, org, repo, baseBranch); err != nil {
+		logger.WithError(err).Warnf("couldn't merge the pull request and push to the destination: %v", err)
+		s.createComment(ic, fmt.Sprintf("Publicize failed with error: %v", err), logger)
+		return
+	}
+
+	bondedOrgRepoLink := fmt.Sprintf("https://github.com/%s", bondedOrgRepo)
+	s.createComment(ic, fmt.Sprintf("Publicize to the repository: [%s/%s](%s) was successful",
+		bondedOrg, bondedRepo, bondedOrgRepoLink), logger)
+}
+
+func (s *server) mergeAndPushToRemote(bondedOrg, bondedRepo string, org, repo, branch string) error {
+	defer func() {
+		if err := s.gc.Clean(); err != nil {
+			logrus.WithError(err).Error("Could not clean up git client cache.")
+		}
+	}()
+
+	repoClient, err := s.gc.ClientFor(bondedOrg, bondedRepo)
+	if err != nil {
+		return fmt.Errorf("couldn't create repoclient for repository %s/%s: %w", bondedOrg, bondedRepo, err)
+	}
+
+	remoteResolver := func() (string, error) {
+		return fmt.Sprintf("https://%s@github.com/%s/%s", string(s.githubTokenGenerator()), org, repo), nil
+	}
+
+	if err := repoClient.FetchFromRemote(remoteResolver, branch); err != nil {
+		return fmt.Errorf("couldn't fetch from bonded repository: %w", err)
+	}
+
+	if err := repoClient.Config("user.name", s.gitName); err != nil {
+		return fmt.Errorf("couldn't set config user.name=%s: %w", s.gitName, err)
+	}
+
+	if err := repoClient.Config("user.email", s.gitEmail); err != nil {
+		return fmt.Errorf("couldn't set config user.name=%s: %w", s.gitEmail, err)
+	}
+
+	mergeMsg := fmt.Sprintf("merge history from %s/%s", bondedOrg, bondedRepo)
+	if merged, err := repoClient.MergeWithStrategy("FETCH_HEAD", "merge", git.MergeOpt{CommitMessage: mergeMsg}); err != nil && !merged {
```

Comment:
I don't think this is correct: at this point you basically did:

```
$ git init (and whatever ClientFor is doing internally)
$ git fetch remote branch
$ git merge (...options...) FETCH_HEAD
```

Which means you will always merge to a default branch (`master`), not to the desired one.

Reply:
Nice catch. Ok we need to checkout to the desired branched before merging, right?

Reply:
Yes, and before that make sure that we fetch the corresponding remote branch, so that the checked out branch has a full history.